### PR TITLE
Add healthz endpoint that checks the vault token

### PR DESCRIPTION
**Pull request by rossf7**
Fixes giantswarm/giantswarm#1558

Adds healthz endpoint that checks the Vault token and API connection are working. To do this it lists the sys mounts. A health endpoint has been added to the Vault API and will be in the next release. We should switch to that once we've upgraded to Vault 0.7.

https://github.com/hashicorp/vault/blob/master/CHANGELOG.md#073-june-7th-2017

### CODE REVIEW

**Comment by rossf7** [comment]
I had problems rebasing the old PR so I did a fresh one. I think this was due to the big vendor update.

@xh3b4sd @kopiczko your comments from #59 are addressed

@JosephSalisbury adding you for info as I added Prometheus support to the healthz endpoint

File: service/healthz/service.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	micrologger "github.com/giantswarm/microkit/logger"
+	vaultapi "github.com/hashicorp/vault/api"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	healthCheckRequests = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "health_check_request_total",
+			Help: "Number of health check requests",
```
Comment (xh3b4sd) [inline service/healthz/service.go]: The help messages are usually sentences. Please add a full stop.

File: service/healthz/service.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	healthCheckRequestTime = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "health_check_request_milliseconds",
+		Help: "Time taken to respond to health check, in milliseconds",
+	})
+)
+
+func init() {
+	prometheus.MustRegister(healthCheckRequests)
+	prometheus.MustRegister(healthCheckRequestTime)
+}
```
Comment (xh3b4sd) [inline service/healthz/service.go]: I would like to put this into a separate file `metrics.go`.

File: service/healthz/service.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		return nil, microerror.MaskAnyf(invalidConfigError, "logger must not be empty")
+	}
+	if config.VaultClient == nil {
+		return nil, microerror.MaskAnyf(invalidConfigError, "vault client must not be empty")
+	}
+
+	newService := &Service{
+		// Dependencies.
+		vaultClient: config.VaultClient,
+		logger:      config.Logger,
```
Comment (xh3b4sd) [inline service/healthz/service.go]: Minor nitpick. Alphabetical order.

File: service/healthz/service.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+// Check implements the health check which lists the mounts for the sys backend.
+// This checks that the operator can connect to the Vault API and the token is
+// valid.
+func (s *Service) Check(ctx context.Context, request Request) (*Response, error) {
+	start := time.Now()
+	defer func() {
+		healthCheckRequestTime.Set(float64(time.Since(start) / time.Millisecond))
+	}()
```
Comment (xh3b4sd) [inline service/healthz/service.go]: We can do `}(time.Now())`.

File: service/healthz/service.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+	sysBackend := s.vaultClient.Sys()
+
+	_, err := sysBackend.ListMounts()
+	if err != nil {
+		healthCheckRequests.WithLabelValues("failed").Inc()
+		return nil, microerror.MaskAny(err)
+	}
+
+	healthCheckRequests.WithLabelValues("successfull").Inc()
```
Comment (xh3b4sd) [inline service/healthz/service.go]: I would like to have constants for this with some godoc describing what it is for. The label is also called `success`. When the actual value for `success` is `failed` it reads weirdly. Maybe the label should be `result`. Uncertain. /cc @JosephSalisbury

File: service/healthz/service.go (hunk 1 above)
Comment (rossf7) [inline service/healthz/service.go]: Fixed

File: service/healthz/service.go (hunk 2 above)
Comment (rossf7) [inline service/healthz/service.go]: Done

File: service/healthz/service.go (hunk 3 above)
Comment (rossf7) [inline service/healthz/service.go]: Fixed, damnit missed that one ;)

File: service/healthz/service.go (hunk 4 above)
Comment (rossf7) [inline service/healthz/service.go]: I left this as to me it's more readable.

File: service/healthz/service.go (hunk 5 above)
Comment (rossf7) [inline service/healthz/service.go]: I got rid of the magic strings and added godoc. But to me the label names make sense. As only one is applied and they can be used for filtering.

Although I managed to spell successful wrong. Fail! :D

**Comment by r7vme** [comment]
Did we check this?

Because with latest Vault bootstrap manifests i'm getting

```
curl -H "X-Vault-Token: G8S_VAULT_TOKEN" http://172.16.238.2:8200/v1/sys/mounts
{"errors":["permission denied"]}
```
```
107     -policy="pki-issue-policy-${CLUSTER_NAME}" \
108     -policy="secret-g8s-sa-sign-key" \
109     -policy="default" \
```

**Comment by xh3b4sd** [comment]
@r7vme `permission denied` errors in Vault usually mean the Vault token being used is not valid and probably expired.

**Pull request merged**